GitUI/HelperDialogs/FormResetAnotherBranch.cs

```diff
@@ -61,9 +62,28 @@ private IGitRef[] GetLocalBranchesWithoutCurrent()
             var currentBranch = Module.GetSelectedBranch();
             var isDetachedHead = currentBranch == DetachedHeadParser.DetachedBranch;
 
-            return Module.GetRefs(false)
+            IGitRef[] refs = Module.GetRefs(false)
                 .Where(r => r.IsHead && (isDetachedHead || r.LocalName != currentBranch))
                 .ToArray();
+
+            var remotes = _revision.Refs.Where(r => r.IsRemote).ToList();
+            if (remotes.Count == 0)
+            {
+                return refs;
+            }
+
+            var refsOrdered = new List<IGitRef>();
+            foreach (var remote in remotes)
+            {
+                IGitRef localBranch = refs.FirstOrDefault(r => r.MergeWith == remote.LocalName && r.TrackingRemote == remote.Remote);
+                if (localBranch != null)
+                {
+                    refsOrdered.Add(localBranch);
+                }
+            }
+
+            refsOrdered.AddRange(refs.Where(r => !refsOrdered.Contains(r)));
+            return refsOrdered.ToArray();
```

Comment:
I think this proposition is much more difficult to read and get the intent just to avoid multiple enumeration in a not performance constraint action. I prefer to avoid it.

I have proposed something else. Please review again...

Reply:
Thanks. It looks better.
There is another solution:

```
.Select(remote => refs.FirstOrDefault(r => r.MergeWith == remote.LocalName && r.TrackingRemote == remote.Remote))
.Where(r => r != null)
```

to prevent enumerations

```
var indexedRemotes = _revision.Refs
    .Where(r => r.IsRemote)
    .ToDictionary(x => $"{x.LocalName ?? string.Empty}{x.Remote ?? string.Empty}", x => x);

if (!indexedRemotes.Any())
{
    return refs.ToArray();
}

var trackedRefs = refs
    .Where(x => indexedRemotes.ContainsKey($"{x.MergeWith ?? string.Empty}{x.TrackingRemote ?? string.Empty}"));
```

Reply:
Dictionary does O(1).

Reply:
I do not see many use cases where there are more than 5 branches for a commit, this is not where optimizations are needed the most. Go withe most readable implementation. (The updated proposal need an explanation as well, not a big difference.)

Reply:
Another angle of view. To help. I do not insist on any change.

Reply:
I took another direction to make it clearer in an added commit (and it is incompatible with your proposition). Tell me what you think about it...

Reply:
Thanks. Good way.
Could we remove?

`.Where(r => r != null);`

with flipping

```
var trackedRefsToDisplayFirst = refs
    .Where(x => remotes.Any(y => x.IsTrackingRemote(y)));
```

Reply:
I did not go into the details of the problem, but it is possible.

```
refs.OrderBy(x => remotes.Any(y => x.IsTrackingRemote(y)))
   .ThenBy(x => x.Name)
```

Reply:
```
var trackedRefsToDisplayFirst = refs
    .OrderBy(x => remotes.Any(y => x.IsTrackingRemote(y)))
    .ThenBy(x => x.Name)
    .ToArray();

return trackedRefsToDisplayFirst;
```

Reply:
The `.ThenBy(x => x.Name)` is not possible because it will change the behavior that is to respect the order defined by the user in the settings.

![image](https://user-images.githubusercontent.com/460196/100524409-3baeff80-31b8-11eb-8fe0-a21bf56a93e8.png)

And if we didn't put this clause, I don't know what is the default behavior. Will it keep the current order?

That's also the reason why I use the `remotes` collection instead of `refs` to iterate because I also wanted to keep that the order they appears. But most of the times their will be only one remote so maybe it's not that important....

Reply:
I ran this code, the order is the same.

```
private IGitRef[] GetLocalBranchesWithoutCurrent()
{
    var currentBranch = Module.GetSelectedBranch();
    var isDetachedHead = currentBranch == DetachedHeadParser.DetachedBranch;
    var remotes = _revision.Refs
        .Where(r => r.IsRemote);

    return Module.GetRefs(tags: false)
        .Where(r => r.IsHead)
        .Where(r => isDetachedHead || r.LocalName != currentBranch)
        .OrderByDescending(x => remotes.Any(y => x.IsTrackingRemote(y)))
        .ToArray();
}
```

Module.GetRefs returns a "List", which should guarantee ordering.

Reply:
![image](https://user-images.githubusercontent.com/3169265/100524913-c72e8d80-31c4-11eb-95be-56a4a64cb960.png)

Reply:
I don't know what is the difference Module.GetRefs and _revision.Refs.

Reply:
refs[].IsRemote != remotes[].IsRemote ?

Reply:
`Module.GetRefs`: all the refs
`_revision.Refs`: the ones displayed in the `_revision` commit line. Here is the selected revision on which the user right clicked to launch the feature.

Updated with what should be the last version. I'm not sure that we can simplify more than that ;)

Reply:
👍